pallets/identity/src/lib.rs

```diff
@@ -1232,106 +1234,125 @@ impl<T: Trait> Module<T> {
+    /// any of CDD providers.
     pub fn has_valid_cdd(claim_for: IdentityId) -> bool {
         let trusted_cdd_providers = T::CddServiceProviders::get_members();
-        Self::is_any_claim_valid(
-            claim_for,
-            IdentityClaimData::CustomerDueDiligence,
-            trusted_cdd_providers,
-        )
+
+        let valid = Self::fetch_claims(claim_for, ClaimType::CustomerDueDiligence, &None)
+            .any(|c| trusted_cdd_providers.contains(&c.claim_issuer));
+
+        valid
     }
 
-    /// IMPORTANT: No state change is allowed in this function
-    /// because this function is used within the RPC calls
-    pub fn is_identity_has_valid_kyc(
-        claim_for: IdentityId,
-        buffer: u64,
-    ) -> (bool, Option<IdentityId>) {
+    /// It returns the CDD identity which issued the current valid CDD claim for `claim_for`
+    /// identity.
+    /// # Parameters
+    ///  - leeway: This leeway is added to now() before check if claim is expired.
+    pub fn fetch_cdd(claim_for: IdentityId, leeway: u64) -> Option<IdentityId> {
         let trusted_cdd_providers = T::CddServiceProviders::get_members();
-        if let Some(threshold) = <pallet_timestamp::Module<T>>::get()
+        let exp_with_leeway = <pallet_timestamp::Module<T>>::get()
             .saturated_into::<u64>()
-            .checked_add(buffer)
-        {
-            for trusted_cdd_provider in trusted_cdd_providers {
-                if let Some(claim) = Self::fetch_valid_claim(
-                    claim_for,
-                    IdentityClaimData::CustomerDueDiligence,
-                    trusted_cdd_provider,
-                ) {
-                    if let Some(claim_expiry) = claim.expiry {
-                        if claim_expiry <= threshold {
-                            return (false, None);
-                        }
-                    }
-                    return (true, Some(trusted_cdd_provider));
-                }
-            }
+            .checked_add(leeway)
+            .unwrap_or_default();
+
+        let cdd = Self::fetch_base_claims(claim_for, ClaimType::CustomerDueDiligence, &None)
+            .filter(|id_claim| {
+                Self::is_identity_claim_not_expired_at(id_claim, exp_with_leeway)
+                    && trusted_cdd_providers.contains(&id_claim.claim_issuer)
+            })
+            .map(|id_claim| id_claim.claim_issuer)
+            .nth(0);
+        cdd
+    }
+
+    /// It iterates over all claims of type `claim_type` for target `id` identity.
+    /// Please note that it could return expired claims.
+    fn fetch_base_claims<'a>(
+        target: IdentityId,
+        claim_type: ClaimType,
+        scope: &'a Option<IdentityId>,
+    ) -> impl Iterator<Item = IdentityClaim> + 'a {
+        let pk = Claim1stKey { target, claim_type };
+
+        <Claims>::iter_prefix(pk).filter(move |c| Self::is_identity_claim_in_scope(c, scope))
```

Comment:
We should not be iterating over claims based on just the Claim1stKey because anyone can issue claims and increase the number of claim under a specific Claim1stKey. This becomes a dos vector where Alice can issue thousands of CDD claim on Bob but since Alice is not a CDD provider, all those claims will be invalid and Bob might have to iterate over all those fake claims to finally get to his valid cdd claim.

Reply:
Only CDD providers are going to be able to produce CDD Claims and adding a claim is going to have a cost, so attacker is going to spend a lot of money, and Bob could just remove all its claims of an specific issuer (attacker) on a cheap and single transaction (TBD).

BTW It sounds quite interesting to limit the case of a rich attacker, e.g:
- The cost of adding a claim based on the size of the number of elements contained in `Claim1stKey`. Let say constant for regular case (< 20 claims), linear incremental for busy target identity ( >= 20 & <50 claims), exponential for attackers( >50 claims).

Again this case seems out of the box of this task and we could manage it on an specific security one.

Reply:
Claim receivers can not remove claims. Only claim issuers can remove claims so bob will never be able to remove Alice's claim.

This "attack" is being introduced by this PR so I think it makes sense to solve within this PR. That being said, I understand that this PR has already been stretched enough and I'll be fine if we create a new task to tackle this problem but it's something we need to tackle before V1 release. cc @adamdossa